File: internal/adminops/honesty_test.go
package adminops_test

import (
	"context"
	"strings"
	"testing"

	"github.com/heros-foreal/agentd/internal/adminaudit"
	"github.com/heros-foreal/agentd/internal/adminops"
	"github.com/heros-foreal/agentd/internal/adminrbac"
	"github.com/heros-foreal/agentd/internal/authoring"
	"github.com/heros-foreal/agentd/internal/linkingest"
)

// honesty_test.go defends the three corrections P26 wave 26a made to figures that were ALREADY
// SHIPPING on the operator console (tasks 2.2, 2.6, 2.7, and their supporting 2.3 and 2.4).
//
// These are not new-feature tests. Each one names the requirement it defends, so a later change that
// removes the correction fails with an explanation rather than with a diff nobody can interpret. That
// matters more here than usual: all three defects survived fourteen phases precisely because nothing
// asserted them, and a correction with no assertion is a correction with a half-life.

// TestLinkCoverageIsDisplayedBesideEverySUMDerivedFigure defends P26 task 2.2 and the requirement
// "Link coverage SHALL be displayed beside every SUM-derived figure on the operator console".
//
// 🔴 If this fails, the operator billing surface has gone back to showing a SUM figure that nothing
// qualifies. `openspec/project.md` states the rule once: metering counts only what it observed, and
// link coverage is displayed wherever a derived figure is shown. A billing operator issuing a credit
// against a SUM figure with 31% coverage is acting on a number wrong by an unknown factor.
func TestLinkCoverageIsDisplayedBesideEverySUMDerivedFigure(t *testing.T) {
	h := newHarness(t)
	h.seedMeteredCharge(tenantAcme)

	// What the platform OBSERVED: ten runs reported, four linked. The figure below reflects four.
	h.links.ObserveRunsReported(tenantAcme, 10)
	for i := 0; i < 4; i++ {
		if _, err := h.links.Record(linkingest.LinkedRun{
			RunID: "run-acme-" + string(rune('a'+i)), TenantID: tenantAcme, LinkedAt: h.clk.now(),
		}); err != nil {
			t.Fatalf("Record: %v", err)
		}
	}

	view, err := h.bills.Oversight(h.ctx(adminrbac.RoleBillingOps), tenantAcme, h.period)
	if err != nil {
		t.Fatalf("Oversight: %v", err)
	}

	if !view.LinkCoverage.Known {
		t.Fatal("link coverage is not known — the surface would show a SUM figure it cannot qualify")
	}
	if got := view.LinkCoverage.Percent; got < 39 || got > 41 {
		t.Fatalf("link coverage = %.1f%%, want 40%% (4 of 10 runs linked)", got)
	}
	if view.MeteredSUM == nil {
		t.Fatal("no SUM-derived figure was produced with known coverage")
	}
	// 🔴 The pairing is in the TYPE. This is the assertion that fails if somebody adds a derived figure
	// without its coverage, because there is no constructor that produces one.
	if view.MeteredSUM.Coverage == nil {
		t.Fatal("the SUM-derived figure carries no coverage — the pairing that makes this correction " +
			"structural has been broken; see adminops.DerivedFigure")
	}
	if *view.MeteredSUM.Coverage != view.LinkCoverage.Percent {
		t.Fatalf("the figure's coverage (%.1f) disagrees with the surface's (%.1f) — two sources for one claim",
			*view.MeteredSUM.Coverage, view.LinkCoverage.Percent)
	}
	// The coverage must be a statement in the same view, not a number the reader has to interpret.
	if !strings.Contains(view.LinkCoverage.Statement, "LINKED runs only") {
		t.Fatalf("the coverage statement does not say what the figure counts: %q", view.LinkCoverage.Statement)
	}
	if !strings.Contains(view.LinkCoverage.Statement, "never inferred") {
		t.Fatalf("the coverage statement does not refuse extrapolation: %q", view.LinkCoverage.Statement)
	}
	// The rows the coverage qualifies are marked, so a percentage beside the table does not appear to
	// qualify a seat count or a plan change.
	var sumDerived int
	for _, line := range view.Invoices {
		if line.SUMDerived {
			sumDerived++
		}
	}
	if sumDerived == 0 {
		t.Fatal("no invoice row is marked SUM-derived — the coverage would appear to qualify every row")
	}
}

// TestAFigureWithUnknownCoverageIsWithheld defends P26 task 2.3 and the requirement "A figure whose
// coverage is unknown SHALL NOT be rendered".
//
// The distinction under test is the one that is invisible on screen: 0% coverage and unknown coverage
// look identical as a number and mean opposite things.
func TestAFigureWithUnknownCoverageIsWithheld(t *testing.T) {
	h := newHarness(t)
	h.seedMeteredCharge(tenantBoreal)
	// Nothing reported, nothing linked: coverage is UNKNOWN, not zero.

	view, err := h.bills.Oversight(h.ctx(adminrbac.RoleBillingOps), tenantBoreal, h.period)
	if err != nil {
		t.Fatalf("Oversight: %v", err)
	}
	if view.LinkCoverage.Known {
		t.Fatal("coverage reported as known with nothing linked and nothing reported")
	}
	if view.MeteredSUM != nil {
		t.Fatalf("a SUM-derived figure was produced with unknown coverage: %+v — the figure must be "+
			"withheld, because a wrong number gets acted on and a missing one prompts a question", *view.MeteredSUM)
	}
	if view.GainshareSavings != nil {
		t.Fatal("a gainshare figure was produced with unknown coverage")
	}
	if !strings.Contains(view.LinkCoverage.Statement, "unknown") {
		t.Fatalf("the surface does not state that coverage is unknown: %q", view.LinkCoverage.Statement)
	}
	// And it must not be filled with an estimate.
	for _, word := range []string{"estimate", "approximately", "projected"} {
		if strings.Contains(strings.ToLower(view.LinkCoverage.Statement), word) {
			t.Fatalf("the unknown-coverage statement offers an estimate (%q): %q", word, view.LinkCoverage.Statement)
		}
	}
}

// TestAGainshareFigureNamesTheVerifiedDeltaLedger defends P26 task 2.4 and the requirement "A
// gainshare or verified-savings figure SHALL name the verified-delta ledger it drew on".
//
// Provenance on the SURFACE, not in a document: an operator has to be able to tell a verified figure
// from an unverified estimate without leaving the page.
func TestAGainshareFigureNamesTheVerifiedDeltaLedger(t *testing.T) {
	h := newHarness(t)
	h.seedMeteredCharge(tenantAcme)
	h.links.ObserveRunsReported(tenantAcme, 2)
	for i := 0; i < 2; i++ {
		if _, err := h.links.Record(linkingest.LinkedRun{
			RunID: "run-acme-g" + string(rune('a'+i)), TenantID: tenantAcme, LinkedAt: h.clk.now(),
		}); err != nil {
			t.Fatalf("Record: %v", err)
		}
	}

	view, err := h.bills.Oversight(h.ctx(adminrbac.RoleBillingOps), tenantAcme, h.period)
	if err != nil {
		t.Fatalf("Oversight: %v", err)
	}
	if view.GainshareSavings == nil {
		t.Fatal("no gainshare figure was produced with complete coverage")
	}
	if view.GainshareSavings.Source != adminops.SourceVerifiedDeltaLedger {
		t.Fatalf("the gainshare figure names %q as its source, want %q — a savings figure drawing on "+
			"anything else is an unverified estimate wearing a billable figure's clothes",
			view.GainshareSavings.Source, adminops.SourceVerifiedDeltaLedger)
	}
	if !strings.Contains(view.GainshareSavings.Basis, "Unverified authored changes are excluded") {
		t.Fatalf("the gainshare figure does not state the exclusion: %q", view.GainshareSavings.Basis)
	}
	// And the SUM figure must NOT claim the verified-delta ledger — two figures, two provenances.
	if view.MeteredSUM != nil && view.MeteredSUM.Source == adminops.SourceVerifiedDeltaLedger {
		t.Fatal("the metered SUM figure claims the verified-delta ledger as its source")
	}
}

// TestASeededUnverifiedAuthoredChangeContributesExactlyZero defends P26 tasks 2.5 and 2.6 and the
// requirement "Every aggregate improvement, savings or quality figure SHALL exclude unverified
// authored changes".
//
// 🔴 It seeds one and reads the FIGURES. Asserting that a `WHERE` clause exists is asserting that we
// wrote a `WHERE` clause; this asserts the number does not move. The difference is the whole point —
// this repository has already found a green suite sitting on top of a fake written to match it.
func TestASeededUnverifiedAuthoredChangeContributesExactlyZero(t *testing.T) {
	h := newHarness(t)
	ctx := h.ctx(adminrbac.RoleBillingOps)

	before, err := h.crossTenant.View(ctx, adminops.AggregateAuthoredImprovement, h.period)
	if err != nil {
		t.Fatalf("View: %v", err)
	}
	baseline := figuresOf(t, before)

	// One authored change, applied and never measured by the harness.
	if err := h.authored.Append(context.Background(), authoring.Entry{
		ChangeID: "chg-unverified-1", Action: authoring.ActionSubmitted, TenantID: tenantAcme,
		ActorID: "user-1", WorkflowID: "wf-1", ConfigHash: "cfg-unverified",
		Axis: "prompt", Origin: "studio", VerificationState: authoring.StateUnverified, At: h.clk.now(),
	}); err != nil {
		t.Fatalf("Append: %v", err)
	}

	after, err := h.crossTenant.View(ctx, adminops.AggregateAuthoredImprovement, h.period)
	if err != nil {
		t.Fatalf("View: %v", err)
	}
	got := figuresOf(t, after)

	for _, label := range adminops.ImprovementFigures {
		if got[label] != baseline[label] {
			t.Fatalf("%s moved from %v to %v after seeding ONE UNVERIFIED authored change. "+
				"An applied-but-unmeasured change must contribute exactly zero to every improvement, "+
				"savings and quality figure (P26 §2.5; authoring.CountableAggregate is the one filter).",
				label, baseline[label], got[label])
		}
	}
	// And the exclusion must be VISIBLE — an invisible exclusion is indistinguishable from an oversight.
	if got[adminops.RowExcludedUnverified] != baseline[adminops.RowExcludedUnverified]+1 {
		t.Fatalf("%s did not count the seeded change (%v → %v) — the surface must be able to say "+
			"'we looked at this and did not count it'", adminops.RowExcludedUnverified,
			baseline[adminops.RowExcludedUnverified], got[adminops.RowExcludedUnverified])
	}
	if !strings.Contains(after.Note, "Unverified authored changes are excluded") {
		t.Fatalf("the aggregate does not state the exclusion where the figures appear: %q", after.Note)
	}

	// The control half: a VERIFIED change DOES move the improvement figure. Without this, a filter that
	// excluded everything would pass the assertion above and the test would prove nothing.
	if err := h.authored.Append(context.Background(), authoring.Entry{
		ChangeID: "chg-verified-1", Action: authoring.ActionSubmitted, TenantID: tenantAcme,
		ActorID: "user-1", WorkflowID: "wf-1", ConfigHash: "cfg-verified",
		Axis: "prompt", Origin: "studio", VerificationState: authoring.StateVerified, At: h.clk.now(),
	}); err != nil {
		t.Fatalf("Append: %v", err)
	}
	verified, err := h.crossTenant.View(ctx, adminops.AggregateAuthoredImprovement, h.period)
	if err != nil {
		t.Fatalf("View: %v", err)
	}
	if figuresOf(t, verified)[adminops.ImprovementFigures[0]] != baseline[adminops.ImprovementFigures[0]]+1 {
		t.Fatal("a VERIFIED authored change did not move the improvement figure — the exclusion is " +
			"excluding everything, and the assertion above would pass on a broken aggregate")
	}
}

// TestTheAuditSurfaceStatesWhichMergePathsTheChainCovers defends P26 task 2.7 and the requirement
// "The audit chain's merge coverage SHALL be stated honestly".
//
// The chain mirrors P6 autonomous merges. It does not, and structurally cannot, mirror P12
// customer-CI-mediated deliveries — those merge in the customer's CI under a credential the platform
// does not hold. An audit log described as the record of "every merge" now covers one of two paths,
// and an auditor concluding "no record, so it did not happen" would be drawing a false conclusion
// from a true page.
func TestTheAuditSurfaceStatesWhichMergePathsTheChainCovers(t *testing.T) {
	h := newHarness(t)

	view, err := h.auditView.Entries(h.ctx(adminrbac.RolePlatformSRE), adminaudit.Filter{}, 10)
	if err != nil {
		t.Fatalf("Entries: %v", err)
	}
	cov := view.MergeCoverage

	if len(cov.Covered) == 0 {
		t.Fatal("the audit surface names no covered merge path")
	}
	if len(cov.NotCovered) == 0 {
		t.Fatal("the audit surface names no UNCOVERED merge path — which is the whole correction: " +
			"before P26 it implied it covered every merge")
	}
	var sawAutonomous, sawCIMediated bool
	for _, p := range cov.Covered {
		if p.ID == "p6-autonomous-merge" {
			sawAutonomous = true
			if !strings.Contains(p.Mechanism, "mergeaudit.go") {
				t.Fatalf("the covered path does not name the code that records it: %q", p.Mechanism)
			}
		}
	}
	for _, p := range cov.NotCovered {
		if p.ID == "p12-ci-mediated-delivery" {
			sawCIMediated = true
			// A gap with a destination is a different thing from a gap.
			if p.ReadableAt != "/delivery" {
				t.Fatalf("the uncovered path does not name where it IS readable: %q", p.ReadableAt)
			}
		}
	}
	if !sawAutonomous || !sawCIMediated {
		t.Fatalf("merge-path coverage is incomplete: autonomous=%v ci-mediated=%v", sawAutonomous, sawCIMediated)
	}
	if !strings.Contains(cov.Statement, "does not") && !strings.Contains(cov.Statement, "NOT") {
		t.Fatalf("the statement does not say what the chain omits: %q", cov.Statement)
	}
	if !strings.Contains(cov.Statement, "not evidence that it did not happen") {
		t.Fatalf("the statement does not warn against reading absence as evidence: %q", cov.Statement)
	}
}

// figuresOf indexes an aggregate's rows by label, failing on a duplicate — two rows with one label
// would make every assertion above depend on iteration order.
func figuresOf(t *testing.T, m adminops.ReadModel) map[string]float64 {
	t.Helper()
	out := map[string]float64{}
	for _, r := range m.Rows {
		if _, dup := out[r.Label]; dup {
			t.Fatalf("aggregate %s has two rows labelled %q", m.Aggregate, r.Label)
		}
		out[r.Label] = r.Value
	}
	return out
}

File: web/admin-console/tests/honesty.test.mjs
// honesty.test.mjs is the console half of P26's three corrections (tasks 2.2, 2.6, 2.7, 2.8).
//
// The platform half lives in `internal/adminops/honesty_test.go`, where a seeded unverified change is
// read back out of the real aggregate. This file asserts the thing the Go side cannot: that the
// corrections REACH THE SCREEN, in the same view, and that a later change cannot quietly render a
// derived figure without its coverage.
//
// Each test names the requirement it defends. A correction with no assertion has a half-life, and all
// three of these defects survived fourteen phases precisely because nothing asserted them.

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFile, readdir, writeFile } from "node:fs/promises";
import { join } from "node:path";

const ROOT = new URL("..", import.meta.url).pathname;

async function read(rel) {
  return readFile(join(ROOT, rel), "utf8");
}

async function pages() {
  const out = [];
  async function walk(dir) {
    for (const entry of await readdir(join(ROOT, dir), { withFileTypes: true })) {
      const rel = join(dir, entry.name);
      if (entry.isDirectory()) await walk(rel);
      else if (rel.endsWith(".tsx")) out.push(rel);
    }
  }
  await walk("src/app");
  return out;
}

// ── 2.2 · Link coverage beside every SUM-derived figure ──────────────────────

test("🔴 2.2 — a SUM-derived figure reaches the screen ONLY through the component that carries its coverage", async () => {
  // The fence, and the reason it is written this way: the requirement is "coverage beside EVERY
  // SUM-derived figure", and the way that requirement fails is by somebody adding a fourth figure and
  // rendering `{view.something.value}` directly. So the assertion is not "the billing page shows
  // coverage" — it is that no page can read a DerivedFigure's value at all except through <Derived>.
  const types = await read("src/lib/types.ts");
  const derivedFields = [...types.matchAll(/(\w+)\??:\s*DerivedFigure;/g)].map((m) => m[1]);
  assert.ok(
    derivedFields.length > 0,
    "no field is typed DerivedFigure — the pairing type has been removed from the wire contract",
  );

  for (const file of await pages()) {
    const src = await read(file);
    for (const field of derivedFields) {
      const direct = new RegExp(`\\.${field}\\.value`);
      assert.equal(
        direct.test(src),
        false,
        `${file} renders ${field}.value directly. A SUM-derived figure must go through <Derived>, ` +
          `which renders its link coverage in the same view — or withholds the figure when coverage ` +
          `is unknown (P26 §2.2, §2.3).`,
      );
    }
  }
});

test("🔴 2.2 — the billing surface states its coverage and renders both derived figures", async () => {
  const page = await read("src/app/billing/page.tsx");
  assert.match(page, /<CoverageStatement coverage={view\.link_coverage}/, "no coverage statement on the billing surface");
  assert.match(page, /figure={view\.metered_sum}/, "the metered SUM figure is not rendered");
  assert.match(page, /figure={view\.gainshare_savings}/, "the verified-savings figure is not rendered");
  // In the SAME view, not behind a link and not on a detail page: the coverage section must appear
  // before the tables it qualifies.
  const coverageAt = page.indexOf("What these figures count");
  const invoicesAt = page.indexOf('title="Invoices"');
  assert.ok(coverageAt > -1 && invoicesAt > -1, "the sections were renamed — this assertion needs updating");
  assert.ok(
    coverageAt < invoicesAt,
    "the coverage statement is rendered after the figures it qualifies — an operator about to act on " +
      "a figure must see the coverage without navigating or scrolling past it",
  );
});

// ── 2.3 · A figure with unknown coverage is not rendered ─────────────────────

test("🔴 2.3 — the derived component withholds the figure when coverage is unknown, and says so", async () => {
  const src = await read("src/components/derived.tsx");
  assert.match(
    src,
    /figure\.coverage === null/,
    "the derived component does not branch on unknown coverage",
  );
  assert.match(src, /derived--withheld/, "there is no withheld rendering — an unknown-coverage figure would vanish");
  assert.match(src, /link coverage is unknown/i, "the withheld rendering does not say why the figure is absent");
  // The withheld state must not be an emptiness. A slot that simply disappears is indistinguishable
  // from a page that failed to load, which is the distinction the nine states exist for.
  const css = await read("src/app/globals.css");
  assert.match(css, /\.derived--withheld\s*{/, "the withheld state has no rendering of its own in the stylesheet");
});

test("🔴 2.3 — the coverage sits at reading size, not at hint size", async () => {
  // A caveat set at hint scale beside a display-scale figure reads as decoration. This one changes
  // whether an operator issues a credit, so it is set at --text-sm in the page ink rather than
  // --text-xs in the muted ink.
  const css = await read("src/app/globals.css");
  const block = css.match(/\.derived__coverage\s*{([^}]*)}/);
  assert.ok(block, "no .derived__coverage rule");
  assert.match(block[1], /font-size:\s*var\(--text-sm\)/, "the coverage is not set at reading size");
  assert.match(block[1], /color:\s*var\(--text\)/, "the coverage is set in the muted ink — it reads as decoration");
});

// ── 2.6 · The exclusion is stated where the figure appears ───────────────────

test("🔴 2.6 — the cross-tenant surface states the unverified exclusion in the same view as the figures", async () => {
  const page = await read("src/app/crosstenant/page.tsx");
  assert.match(page, /authored_improvement/, "the improvement aggregate is not reachable from the console");
  assert.match(
    page,
    /model\.note/,
    "the aggregate's exclusion note is not rendered — an operator cannot tell a measured improvement " +
      "from an unverified estimate without leaving the page (P26 §2.6)",
  );
});

// ── 2.7 · The audit surface states its merge-path coverage ───────────────────

test("🔴 2.7 — the audit surface states which merge paths the chain covers and which it does not", async () => {
  const page = await read("src/app/audit/page.tsx");
  assert.match(page, /merge_coverage\.statement/, "the merge-coverage statement is not rendered");
  assert.match(page, /merge_coverage\.covered/, "the covered merge paths are not rendered");
  assert.match(
    page,
    /merge_coverage\.not_covered/,
    "the UNCOVERED merge paths are not rendered — which is the whole correction. Before P26 the " +
      "surface implied the chain recorded every merge, while it records the ones the P6 loop performs " +
      "itself (P26 §2.7).",
  );
  // A gap with a destination is a different thing from a gap.
  assert.match(page, /readable_at/, "an uncovered path does not link to where it IS readable");
  // And the page's own lede must not keep claiming what it no longer claims.
  const lede = page.match(/lede="([^"]*)"/);
  assert.ok(lede, "the audit page has no lede");
  assert.match(
    lede[1],
    /AUTONOMOUS merge|does not record every merge/,
    `the audit page's lede still implies it holds every merge: ${lede[1]}`,
  );
});

// ── 8.5 · The derived-figure fence, demonstrated RED ─────────────────────────

test("🔴 8.5 — the derived-figure fence goes RED when a page renders a figure without its coverage", async () => {
  // A fence nobody has seen fail is a fence nobody knows is connected. This commits the exact
  // violation — a page reading `.value` off a DerivedFigure directly, bypassing <Derived> and its
  // coverage — and requires the assertion above to catch it. The file is restored byte-for-byte.
  const path = join(ROOT, "src/app/billing/page.tsx");
  const original = await readFile(path, "utf8");
  const marker = "<CoverageStatement coverage={view.link_coverage} />";
  assert.ok(original.includes(marker), "the billing page changed shape — this demonstration needs updating");
  try {
    await writeFile(path, original.replace(marker, "<p>{view.metered_sum.value}</p>"));

    const types = await read("src/lib/types.ts");
    const derivedFields = [...types.matchAll(/(\w+)\??:\s*DerivedFigure;/g)].map((m) => m[1]);
    const patched = await readFile(path, "utf8");
    const caught = derivedFields.some((f) => new RegExp(`\\.${f}\\.value`).test(patched));
    assert.ok(
      caught,
      "the fence did NOT catch a SUM-derived figure rendered without its coverage. Its whole purpose " +
        "is that a fourth figure cannot be added bare (P26 §2.2).",
    );
  } finally {
    await writeFile(path, original);
  }
});
